Implement docker

adelefelicia opened the pull request:
- Implement support for running app in Docker containers
- To get the app to work both with and without Docker, I made all calls to backend refer to a API url const set in the env file
- Improved some redundant url references in frontend (defined base url for api instead of repeating for every function)
  - However, I did not change the ones that did already have base urls but that could be improved further ie. base url not including /nodes/ when the whole file uses that endpoint anyways, because I didn't want to change too much frontend in a backend PR
- Commented out problematic call to device connection end point that is not implemented yet / for which the implementation will change

To test this PR:
- Set up app in docker by opening docker desktop (or starting docker via a terminal) and run "docker-compose up --build" in the root of the project
- Some functionality will not work if you open the frontend bc it is yet to be fixes by the frontend team, but you should be able to create and run a test flow and open the logs

EliHaugu commented on docker-compose.yml:
```diff
@@ -0,0 +1,48 @@
+    build:
+      context: .
+      dockerfile: client/Dockerfile
+    ports:
```
is there a reason for making a volume connecting ./server to /app but not ./client?

adelefelicia replied:
No, just forgot. Will add it!

EliHaugu commented on docker-compose.yml:
```diff
@@ -0,0 +1,48 @@
+  client:
+    restart: always
+    build:
+      context: .
```
why is the context for server "./server" but the context for client is "./client"?

adelefelicia replied:
Was wondering this myself, but didn't get around to investigating. Looked into it now and seems they can both have the context as their respective directories. I think it was different because different people made the initial Docker setup for client and server

EliHaugu commented on docker-compose.yml:
```diff
@@ -0,0 +1,48 @@
+      - "5173:5173"
+    networks:
+      - app_network
+    command: npm run dev -- --host
```
do we need this command both in this file and the client Dockerfile?

adelefelicia replied:
No, we don't. Good catch. I'll delete it from this file.

adelefelicia commented:
Added some things to make it run better:
- Health check that db is running before server tries to run
- Added volume for node modules (this causes them to have to be installed everytime the containers are run, but I didn't find a better fix)
- Made a vite dependency optional (seems like a VITE bug that interferes with Docker setup https://github.com/vitejs/vite/discussions/15532 )
- Changed BASE_DIR back to parent.parent instead of parent x4 since it is used elsewhere with the former definition in mind
- Added variable for locating root dir for .env file location in django server

EliHaugu commented:
Oooo really great job!! <3

EliHaugu commented on server/Dockerfile:
```diff
@@ -0,0 +1,22 @@
+COPY requirements.txt /app/
+RUN pip install --no-cache-dir -r requirements.txt
+
+COPY . /app
```
ohh gotcha, makes sense, thank u! <3

EliHaugu commented on client/package.json:
```diff
@@ -46,5 +46,8 @@
-  "engineStrict": true
+  "engineStrict": true,
+  "optionalDependencies": {
+    "@rollup/rollup-linux-x64-gnu": "^4.26.0"
```
can this potentially cause issues or nah @Furdook ?

EliHaugu commented on server/server_comm/test_runner/views.py:
```diff
@@ -37,12 +37,13 @@ def post(self, request, flow_id=None):
-                    flow_result["error"] = res["message"]
-                    results.append(flow_result)
-                    break
+                # TODO comment back in when this endpoint is implemented
```
isn't this endpoint done? :o

adelefelicia replied:
Was not done when the branch was created, will remove comment:)
